step-templates/octopus-delete-machine-registration.json

```diff
@@ -0,0 +1,65 @@
+{
+    "Id": "e4255fcb-fe7d-4d5b-8ec0-0243e5f48a9c",
+    "Name": "Delete Target or Worker Registration From Octopus",
+    "Description": "Step that will attempt to delete a target or worker from Octopus Deploy using the API.  If it cannot delete the target or worker it will disable the target and rename it.",
+    "ActionType": "Octopus.Script",
+    "Version": 1,
+    "Author": "octopusbob",
+    "Packages": [],
+    "Properties": {
+      "Octopus.Action.RunOnServer": "true",
+      "Octopus.Action.Script.ScriptSource": "Inline",
+      "Octopus.Action.Script.Syntax": "PowerShell",
+      "Octopus.Action.Script.ScriptBody": "[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12\n\n$OctopusAPIKey = $OctopusParameters[\"DeleteTarget.Octopus.Api.Key\"]\n$TargetName = $OctopusParameters[\"DeleteTarget.Target.Name\"]\n$OctopusUrl = $OctopusParameters[\"DeleteTarget.Octopus.Base.Url\"]\n$SpaceId = $OctopusParameters[\"Octopus.Space.Id\"]\n$TargetType = $OctopusParameters[\"DeleteTarget.Target.TargetType\"]\n\nWrite-Host \"Target Name: $TargetName\"\nWrite-Host \"Octopus URL: $OctopusUrl\"\nWrite-Host \"Space Id: $SpaceId\"\nWrite-Host \"Target Type: $TargetType\"\n\n$header = New-Object \"System.Collections.Generic.Dictionary[[String],[String]]\"\n$header.Add(\"X-Octopus-ApiKey\", $OctopusAPIKey)\n\n$baseTargetUrl = \"$OctopusUrl/api/$SpaceId/machines\"\n\nif ($TargetType -eq \"Worker\")\n{\n\t$baseTargetUrl = \"$OctopusUrl/api/$SpaceId/workers\"\n    Write-Host \"Worker was selected, switching over to use the URL $baseTargetUrl\"\n}\n\n$targetListUrl = \"$($baseTargetUrl)?skip=0&take=1000&partialName=$TargetName\"\nWrite-Host \"Get a list of all machine using the URL $targetListUrl\"\n\n$targetList = (Invoke-RestMethod $targetListUrl -Headers $header)\n\nforeach($target in $targetList.Items)\n{\n    if ($target.Name -eq $TargetName)\n    {\n        $targetId = $target.Id\n        $itemToDeleteEndPoint = \"$baseTargetUrl/$targetId\"\n        try\n        {        \t\n        \tWrite-Highlight \"Deleting the machine $targetId because the name $($target.Name) matches the $TargetName $itemToDeleteEndPoint\"\n        \t$deleteResponse = (Invoke-RestMethod $itemToDeleteEndPoint -Headers $header -Method Delete)\n            Write-Highlight \"Successfully deleted machine $TargetName\"\n            Write-Host \"Delete Response $deleteResponse\"\n        }\n        catch\n        {\n        \t$currentDate = Get-Date -Format \"_MMddyyyy_HHmm\"\n        \t$target.Name = \"$($target.Name)-old$currentdate\"\n            $target.IsDisabled = $True\n            \n            $jsonRequest = $target | ConvertTo-Json\n                        \n            Write-Highlight \"There was an error deleting the machine, renaming it to $($target.name) and disabling it\"\n            $machineResponse = Invoke-RestMethod $itemToDeleteEndPoint -Headers $header -Method PUT -Body $jsonRequest\n        } \n        \n        break\n    }\n}"
```

Comment:
Maybe not a big concern, but this will fail on pre-Spaces version (and if the default space is disabled I think).

Other steps have used logic like this to check for Spaces:
```
function Invoke-OctopusApi {
    param(
        [Parameter(Position = 0, Mandatory)]$Uri,
        [ValidateSet("Get", "Post", "Put", "Delete")]$Method = 'Get',
        $Body
    )
    
    $uriParts = @($octopusBaseUrl, $Uri.TrimStart('/'))    
    $uri = ($uriParts -join '/')

    Write-Verbose "Uri: $uri"
    
    $requestParameters = @{
        Uri = $uri
        Method = $Method
        Headers = @{ "X-Octopus-ApiKey" = $apiKey }
        UseBasicParsing = $true
    }
    
    if ($null -ne $Body) { $requestParameters.Add('Body', ($Body | ConvertTo-Json -Depth 10)) }
    
    return Invoke-WebRequest @requestParameters | % Content | ConvertFrom-Json
}

function Test-SpacesApi {
	Write-Verbose "Checking API compatibility";
	$rootDocument = Invoke-OctopusApi 'api/';
    if($rootDocument.Links -ne $null -and $rootDocument.Links.Spaces -ne $null) {
    	Write-Verbose "Spaces API found"
    	return $true;
    }
    Write-Verbose "Pre-spaces API found"
    return $false;
}

if([string]::IsNullOrWhiteSpace($spaceId)) {
	if(Test-SpacesApi) {
      	$spaceId = $OctopusParameters['Octopus.Space.Id'];
      	if([string]::IsNullOrWhiteSpace($spaceId)) {
          	throw "This step needs to be run in a context that provides a value for the 'Octopus.Space.Id' system variable. In this case, we received a blank value, which isn't expected - please reach out to our support team at https://help.octopus.com if you encounter this error or try providing the Space Id parameter.";
      	}
	}
}

$apiPrefix = "api/"

if ($spaceId) {
	$apiPrefix += $spaceId
}
```

Reply:
Invoke-RestMethod you monster

Reply:
![Some men just want to watch the world burn](https://media.giphy.com/media/qkJJRL9Sz1R04/giphy.gif)

Reply:
Got the check down to a few lines of code.